cmd/snap-update-ns/utils.go

```diff
@@ -239,7 +239,7 @@ func MkFile(dirFd int, dirName string, name string, perm os.FileMode, uid sys.Us
 	// we know if we need to chown it) but fall back to just opening an
 	// existing one.
 
-	newFd, err := sysOpenat(dirFd, name, openFlags|syscall.O_CREAT|syscall.O_EXCL, uint32(perm.Perm()))
+	newFd, err := sysOpenat(dirFd, name, openFlags|syscall.O_CREAT|syscall.O_EXCL, uint32(perm))
```

Comment:
I don't think we expect any other bits than the permission bits to be set for files, so it's probably to stay with `perm.Perm()` and leave a comment

Reply:
If we're going to keep any of the `.Perm()` calls, it'd be good to determine what their purpose is. Looking at the revision history, they were added in commit 08d3ed6 but the message doesn't shed much light.

If it is intended as input sanitisation, it seems to be the wrong place (I'd expect it to be done closer to where the input is read). As is, it seems like something that will either trip up future developers or hide bugs (e.g. if someone uses a decimal constant as a mode instead of octal, it'll strip off all the high bits that might help highlight the problem).

Reply:
This is the code the os.Mkdir, os.OpenFile, etc use to convert the os.FileMode type to the form the syscall expects:

https://github.com/golang/go/blob/24b570354caee33d4fb3934ce7ef1cc97fb403fd/src/os/file_posix.go#L61-L75

In particular: it isn't assuming that bits 9-11 are sticky, setuid, setgid. In fact Go's `os.FileMode` type uses a different set of bits for those flags.

So I think there are two possible ways forward:

1. stop pretending that these are os.FileMode values and just use uint32 directly.
2. replace the `.Perm()` calls with calls to a copy of the `syscallMode` helper, and change the AddModeHint() calls from using `01777` to `0777 | os.ModeSticky`.

Reply:
I think it'd go with option 2, since we've basically issuing syscalls ourselves instead of going through `os`, it sounds reasonable to go with the same approach as go stdlib for clarity. Maybe add a reference to the relevant code in `os` in the comment as well.